nemo/collections/nlp/models/language_modeling/megatron_t5_model.py

```diff
@@ -0,0 +1,477 @@
+            )
+        else:
+            raise Exception('{} dataloader type is not supported.'.format(self.cfg.dataloader_type))
+
+        # Torch dataloader.
+        return torch.utils.data.DataLoader(
+            dataset, batch_sampler=batch_sampler, num_workers=self.cfg.data.num_workers, pin_memory=True,
+        )
+
+    def setup(self, stage=None):
+        """A PTL method to setup the training, validation and test datasets."""
+        if stage == 'predict':
+            return
+        if self._train_dl is not None and self._validation_dl is not None:
+            return
+        self.build_train_valid_test_datasets()
+        self.setup_training_data(self.cfg.data)
+        self.setup_validation_data(self.cfg.data)
+        self.setup_test_data(self.cfg.data)
+
+    def setup_training_data(self, cfg):
+        if hasattr(self, '_train_ds'):
+            resume_checkpoint_path = self.trainer.checkpoint_connector.resume_checkpoint_path
+            if resume_checkpoint_path:
+                consumed_samples = int(
+                    float(re.findall(r"consumed_samples\=([0-9]+.[0-9]+)", resume_checkpoint_path)[0])
+                )
+            else:
+                consumed_samples = 0
+            self._train_dl = self.build_pretraining_data_loader(self._train_ds, consumed_samples)
+
+    def setup_validation_data(self, cfg):
+        if hasattr(self, '_validation_ds'):
+            consumed_samples = 0
+            self._validation_dl = self.build_pretraining_data_loader(self._validation_ds, consumed_samples)
+
+    def setup_test_data(self, cfg):
+        if hasattr(self, '_test_ds'):
+            consumed_samples = 0
+            self._test_dl = self.build_pretraining_data_loader(self._test_ds, consumed_samples)
+
+    def compute_consumed_samples(self, global_step):
+        app_state = AppState()
+        consumed_samples = (
+            global_step
+            * app_state.data_parallel_size
+            * self.cfg.micro_batch_size
+            * self.trainer.accumulate_grad_batches
+        )
+        return int(consumed_samples)
+
+    def configure_gradient_clipping(self, *args, **kwargs):
+        """PTL hook to configure gradients.
+           We use gradient clipping implementation from megatron-lm.
+        """
+        clip_val = self.trainer.gradient_clip_val
+        if clip_val is None:
+            return
+
+        clip_val = float(clip_val)
+        if clip_val <= 0:
+            return
+
+        parameters = self.model.parameters()
+        clip_grad_norm_fp32(parameters=parameters, max_norm=clip_val)
+
+    def predict_step(self, batch: Any, batch_idx: int, dataloader_idx: Optional[int] = None) -> Any:
+        request = batch
+        response = self.complete(request)
+        logging.info(f"response: {response}")
+        return response
+
+    def make_inference_attention_mask_3d(self, source_block, target_block, pad_id):
```

Comment:
Can we consolidate all *_attention_mask_* methods? 
The attention mask creation has been duplicated in multiple places.

Reply:
Yeah agreed. I think this may be a good thing to do in the refactor?

Reply:
Also added a docstring for reduced loss.

Reply:
Agreed